Source/DafnyCompilers/CompilerRewriter.dfy

```diff
@@ -0,0 +1,2215 @@
+        case (Set(st), Set(st')) =>
+          && st == st'
+        case (Closure(ctx, vars, body), Closure(ctx', vars', body')) =>
+          EqValue_Closure(v, v')
+
+        // DISCUSS: Better way to write this?  Need exhaustivity checking
+        case (Bool(b), _) => false
+        case (Char(c), _) => false
+        case (Int(i), _) => false
+        case (Real(r), _) => false
+        case (BigOrdinal(o), _) => false
+        case (BitVector(width, value), _) => false
+        case (Map(m), _) => false
+        case (Multiset(ms), _) => false
+        case (Seq(sq), _) => false
+        case (Set(st), _) => false
+        case (Closure(ctx, vars, body), _) => false
+      }
+    }
+
+    predicate {:verify true} {:opaque} EqValue_Closure(v: WV, v': WV)
+      requires v.Closure?
+      requires v'.Closure?
+      decreases ValueTypeHeight(v) + ValueTypeHeight(v'), 0
+    // Equivalence between values: closure case.
+    //
+    // See ``EqValue``.
+    //
+    // Rk.: contrary to ``EqValue``, it seems ok to make ``EqValue_Closure`` opaque.
+    {
+      var Closure(ctx, vars, body) := v;
+      var Closure(ctx', vars', body') := v';
+      && |vars| == |vars'|
+      && (
+      forall fuel:nat, argvs: seq<WV>, argvs': seq<WV> |
+        && |argvs| == |argvs'| == |vars| // no partial applications are allowed in Dafny
+        // We need the argument types to be smaller than the closure types, to prove termination.\
+        // In effect, the arguments types should be given by the closure's input types.
+        && (forall i | 0 <= i < |vars| :: ValueTypeHeight(argvs[i]) < ValueTypeHeight(v))
+        && (forall i | 0 <= i < |vars| :: ValueTypeHeight(argvs'[i]) < ValueTypeHeight(v'))
+        && (forall i | 0 <= i < |vars| :: EqValue(argvs[i], argvs'[i])) ::
+        var res := InterpCallFunctionBody(v, fuel, argvs);
+        var res' := InterpCallFunctionBody(v', fuel, argvs');
+        // We can't use naked functions in recursive setting: this forces us to write the expanded
+        // match rather than using an auxiliary function like `EqPureInterpResult`.
+        match (res, res') {
+          case (Success(ov), Success(ov')) =>
+            // We need to assume those assertions to prove termination: the value returned by a closure
+            // has a type which is smaller than the closure type (its type is given by the closure return
+            // type)
+            assume ValueTypeHeight(ov) < ValueTypeHeight(v);
+            assume ValueTypeHeight(ov') < ValueTypeHeight(v');
+            EqValue(ov, ov')
+          case (Failure(_), Failure(_)) =>
+            true
+          case _ =>
+            false
+        })
+    }
+
+    predicate {:verify true} EqState(ctx: State, ctx': State)
+    {
+      GEqState(EqValue, ctx, ctx')
+    }
+
+    predicate {:verify true} EqCtx(ctx: Context, ctx': Context)
+      requires WellFormedContext(ctx)
+      requires WellFormedContext(ctx')
```

Comment:
Do you use non-wellformed contexts in this file?  If not, we could just create `type Context = c: Interp.Context | WellFormedContext(c)`